pkg/helper/webconsole.go

```diff
@@ -0,0 +1,120 @@
+package helper
+
+import (
+	"context"
+	"fmt"
+	"github.com/RHsyseng/operator-utils/pkg/logs"
+	"github.com/RHsyseng/operator-utils/pkg/resource/read"
+	"github.com/RHsyseng/operator-utils/pkg/utils/kubernetes"
+	consolev1 "github.com/openshift/api/console/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strings"
+	"time"
+)
+
+var logger = logs.GetLogger("openshift-webconsole")
+
+// ConsoleLinkText is the text of the consoleLink shown on the webconsole
+const ConsoleLinkText = "APIManager - 3scale"
+// ConsoleLinkNamePrefix is the prefix applied to Console link for system master
+const ConsoleLinkNamePrefix = "system-master-link"
+
+// ConsoleLinkSupported checks if a ConsoleLink CRD exists
+func ConsoleLinkSupported() error {
+	gvk := schema.GroupVersionKind{Group: "console.openshift.io", Version: "v1", Kind: "ConsoleLink"}
+	return kubernetes.CustomResourceDefinitionExists(gvk)
+}
+
+// ReconcileConsoleLink creates a ConsoleLink object if it doesn't already exist
+func ReconcileConsoleLink(ctx context.Context, c client.Client, req *reconcile.Request) {
+	route := getRoute(ctx, c, req)
+	if route == nil {
+		removeConsoleLink(ctx, c, req)
+	} else {
+		consoleLinkName := fmt.Sprintf("%s-%s", ConsoleLinkNamePrefix, req.Namespace)
+		consoleLink := &consolev1.ConsoleLink{}
+		err := c.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
+		if err != nil && apierrors.IsNotFound(err) {
+			consoleLink = getConsoleLink(consoleLinkName, route, req)
+			if err := c.Create(ctx, consoleLink); err != nil {
+				logger.Error(err, "Console link is not created.")
+			} else {
+				logger.Info("Console link has been created:", consoleLinkName)
+			}
+		} else if err == nil && consoleLink != nil {
+			updateConsoleLink(ctx, route, consoleLink, c)
+		}
+	}
+}
+func getRoute(ctx context.Context, c client.Client, req *reconcile.Request) *routev1.Route {
+	// here the loop tries the best to collect a route, the reconcile event may comes
+	// before the route resources get created/updated
```

Comment:
I'm not sure why this loop is needed.

Why would the reconcile event come before the route gets created/updated?
In the controller code there's a watch for Route resources. An event will be triggered when a route is created/updated/deleted

Reply:
I can't reproduce the case I saw before.  
I removed the loop.

Reply:
:+1: